### ISSUE TITLE
repmgr error

### ISSUE THREAD

**Issue by cloud-arrow** [bitnami/bitnami-docker-postgresql-repmgr#17]
repmgr 19:21:10.63 DEBUG ==> Host 'postgresql-postgresql-ha-postgresql-0.postgresql-postgresql-ha-postgresql-headless.postgresql.svc.cluster.local:5432' is not accessible

**Comment by javsalgar** [comment]
Hi,

Thank you for using Bitnami. Could you provide more information about the steps to reproduce the issue? Which platform are you using? Note that the charts get automatically tested in all the major Kubernetes platforms and their default PV providers.

**Comment by patrickleet** [comment]
Run the chart with preemptible nodes and wait - it will happen

**Comment by patrickleet** [comment]
Node 0 says waiting for Node 1:

```
k logs -n jx-production jx-postgresql-ha-postgresql-0 postgresql
postgresql-repmgr 16:30:09.85 
postgresql-repmgr 16:30:09.85 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 16:30:09.85 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 16:30:09.86 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 16:30:09.86 
postgresql-repmgr 16:30:09.93 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
repmgr 16:30:10.05 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql 16:30:10.06 INFO  ==> Validating settings in POSTGRESQL_* env vars..
repmgr 16:30:10.06 INFO  ==> Querying all partner nodes for common upstream node...
repmgr 16:30:10.18 INFO  ==> There are no nodes with primary role. Assuming the primary role...
repmgr 16:30:10.23 INFO  ==> Preparing PostgreSQL configuration...
postgresql 16:30:10.24 INFO  ==> postgresql.conf file not detected. Generating it...
repmgr 16:30:10.35 INFO  ==> Preparing repmgr configuration...
repmgr 16:30:10.36 INFO  ==> Initializing Repmgr...
postgresql 16:30:10.36 INFO  ==> Initializing PostgreSQL database...
postgresql 16:30:10.36 INFO  ==> Cleaning stale /bitnami/postgresql/data/postmaster.pid file
postgresql 16:30:10.37 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql 16:30:10.43 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql 16:30:10.45 INFO  ==> Deploying PostgreSQL with persisted data...
postgresql 16:30:10.47 INFO  ==> Stopping PostgreSQL...
postgresql-repmgr 16:30:10.53 INFO  ==> ** PostgreSQL with Replication Manager setup finished! **

postgresql 16:30:10.65 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 16:30:10.98 INFO  ==> ** Starting repmgrd **
[2020-05-19 16:30:10] [NOTICE] repmgrd (repmgrd 5.1.0) starting up
INFO:  set_repmgrd_pid(): provided pidfile is /opt/bitnami/repmgr/tmp/repmgr.pid
[2020-05-19 16:30:11] [NOTICE] starting monitoring of node "jx-postgresql-ha-postgresql-0" (ID: 1000)
[2020-05-19 16:30:11] [ERROR] connection to database failed
[2020-05-19 16:30:11] [DETAIL] 
could not translate host name "jx-postgresql-ha-postgresql-1.jx-postgresql-ha-postgresql-headless.jx-production.svc.cluster.local" to address: Name or service not known

[2020-05-19 16:30:11] [DETAIL] attempted to connect using:
  user=cldntvrepmgr password=9JqA3Yyaqa7y4ZYh4f7 connect_timeout=5 dbname=repmgr host=jx-postgresql-ha-postgresql-1.jx-postgresql-ha-postgresql-headless.jx-production.svc.cluster.local port=5432 fallback_application_name=repmgr
[2020-05-19 16:30:11] [ERROR] unable connect to upstream node (ID: 1001), terminating
[2020-05-19 16:30:11] [HINT] upstream node must be running before repmgrd can start
```

Node 1 says waiting for primary node:

```
k logs -n jx-production jx-postgresql-ha-postgresql-1 postgresql 
postgresql-repmgr 16:32:48.98 
postgresql-repmgr 16:32:49.04 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 16:32:49.04 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 16:32:49.04 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 16:32:49.04 
postgresql-repmgr 16:32:49.06 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
repmgr 16:32:49.18 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql 16:32:49.18 INFO  ==> Validating settings in POSTGRESQL_* env vars..
repmgr 16:32:49.24 INFO  ==> Querying all partner nodes for common upstream node...
repmgr 16:32:49.36 INFO  ==> Preparing PostgreSQL configuration...
postgresql 16:32:49.37 INFO  ==> postgresql.conf file not detected. Generating it...
repmgr 16:32:49.46 INFO  ==> Preparing repmgr configuration...
repmgr 16:32:49.47 INFO  ==> Initializing Repmgr...
repmgr 16:32:49.47 INFO  ==> Waiting for primary node...
postgresql 16:33:59.79 INFO  ==> Stopping PostgreSQL...
```

They are both waiting for each other to come up

**Comment by javsalgar** [comment]
Hi,

Which is the platform you are using? Because we test our charts with all major Kubernetes distributions and this incident does not occur. I would like to know more details about your environment so I can reproduce.

**Comment by patrickleet** [comment]
GKE with a preemptible node pool via the bitnami/postgresql-ha chart

```
- name: postgresql-ha	
  repository: https://charts.bitnami.com/bitnami	
  version: 3.2.3 
```

I did configure a bunch of affinity settings and qos and etc, but it happens without any of that

**Comment by javsalgar** [comment]
Hi,

Here it says

```
postgresql 16:30:10.45 INFO  ==> Deploying PostgreSQL with persisted data...
```

Does the same issue happen when there is no persisted data, when deploying from scratch?

**Comment by patrickleet** [comment]
It usually works for a few days and then breaks - perhaps master fails, begins failover, and then node that was used as failover fails before it fully takes over?

**Comment by patrickleet** [comment]
I have not tried it without persisted data - it was an ssd drive though with policy "retain"

**Comment by javsalgar** [comment]
Ok, after re-checking the logs again, it seems an issue when another node is not ready, and therefore the headless service is unable to resolve. Maybe ti makes sense to change the headless service policy with publishNotReadyAddresses to true. @rafariossaa thoughts on that?

**Comment by nungster** [comment]
I am experiencing a similar issue running in Azure.  node 0 comes up fine but node 1 is waiting for primary.

kl -n compliance nungpostgres-postgresql-ha-postgresql-0 postgresql
postgresql-repmgr 20:54:29.65
postgresql-repmgr 20:54:29.65 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 20:54:29.66 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 20:54:29.66 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 20:54:29.67
postgresql-repmgr 20:54:29.68 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
repmgr 20:54:29.77 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql 20:54:29.78 INFO  ==> Validating settings in POSTGRESQL_* env vars..
repmgr 20:54:29.78 INFO  ==> Querying all partner nodes for common upstream node...
repmgr 20:54:34.82 INFO  ==> There are no nodes with primary role. Assuming the primary role...
repmgr 20:54:34.83 INFO  ==> Preparing PostgreSQL configuration...
postgresql 20:54:34.83 INFO  ==> postgresql.conf file not detected. Generating it...
repmgr 20:54:34.88 INFO  ==> Preparing repmgr configuration...
repmgr 20:54:34.88 INFO  ==> Initializing Repmgr...
postgresql 20:54:34.88 INFO  ==> Initializing PostgreSQL database...
postgresql 20:54:34.89 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql 20:54:34.89 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql 20:54:37.73 INFO  ==> Starting PostgreSQL in background...
postgresql 20:54:38.62 INFO  ==> Changing password of postgres
postgresql 20:54:38.65 INFO  ==> Starting PostgreSQL in background...
repmgr 20:54:38.65 INFO  ==> Creating repmgr user: repmgr
repmgr 20:54:38.69 INFO  ==> Creating repmgr database: repmgr
postgresql 20:54:39.33 INFO  ==> Stopping PostgreSQL...
postgresql 20:54:40.33 INFO  ==> Starting PostgreSQL in background...
repmgr 20:54:40.48 INFO  ==> Registering Primary...
postgresql 20:54:40.69 INFO  ==> Loading custom scripts...
postgresql 20:54:40.69 INFO  ==> Stopping PostgreSQL...
postgresql-repmgr 20:54:41.70 INFO  ==> ** PostgreSQL with Replication Manager setup finished! **

postgresql 20:54:41.75 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 20:54:41.97 INFO  ==> ** Starting repmgrd **
[2020-06-01 20:54:41] [NOTICE] repmgrd (repmgrd 5.1.0) starting up
INFO:  set_repmgrd_pid(): provided pidfile is /opt/bitnami/repmgr/tmp/repmgr.pid
[2020-06-01 20:54:42] [NOTICE] starting monitoring of node "nungpostgres-postgresql-ha-postgresql-0" (ID: 1000)
[2020-06-01 20:54:42] [NOTICE] monitoring cluster primary "nungpostgres-postgresql-ha-postgresql-0" (ID: 1000)

devk8s/yaml/postgres on  master [?]
✦ ❯ kl -n compliance nungpostgres-postgresql-ha-postgresql-1 postgresql
postgresql-repmgr 19:24:59.77
postgresql-repmgr 19:24:59.78 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 19:24:59.78 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 19:24:59.78 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 19:24:59.78
postgresql 19:24:59.79 DEBUG ==> Configuring libnss_wrapper...
postgresql-repmgr 19:24:59.80 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
repmgr 19:24:59.89 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql 19:24:59.89 INFO  ==> Validating settings in POSTGRESQL_* env vars..
repmgr 19:24:59.90 INFO  ==> Querying all partner nodes for common upstream node...
repmgr 19:24:59.90 DEBUG ==> Checking node 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432'...
psql: timeout expired
repmgr 19:25:09.92 DEBUG ==> Skipping: failed to get primary from the node 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432'!
repmgr 19:25:09.92 DEBUG ==> Checking node 'nungpostgres-postgresql-ha-postgresql-1.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432'...
psql: could not connect to server: Connection refused
	Is the server running on host "nungpostgres-postgresql-ha-postgresql-1.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local" (10.120.84.203) and accepting
	TCP/IP connections on port 5432?
repmgr 19:25:09.94 DEBUG ==> Skipping: failed to get primary from the node 'nungpostgres-postgresql-ha-postgresql-1.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432'!
repmgr 19:25:09.94 DEBUG ==> Primary node: 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432'
repmgr 19:25:09.95 INFO  ==> Preparing PostgreSQL configuration...
repmgr 19:25:09.95 DEBUG ==> Injecting a new postgresql.conf file...
postgresql 19:25:09.96 INFO  ==> postgresql.conf file not detected. Generating it...
repmgr 19:25:10.03 DEBUG ==> Injecting a new pg_hba.conf file...
repmgr 19:25:10.03 INFO  ==> Preparing repmgr configuration...
repmgr 19:25:10.04 DEBUG ==> Node ID: '1001', Rol: 'standby', Primary Node: 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432'
repmgr 19:25:10.04 INFO  ==> Initializing Repmgr...
repmgr 19:25:10.04 INFO  ==> Waiting for primary node...
repmgr 19:25:10.05 DEBUG ==> Wait for schema repmgr.repmgr on 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432', will try 6 times with 10 delay seconds (TIMEOUT=60)
psql: timeout expired
repmgr 19:25:20.06 DEBUG ==> Host 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432' is not accessible
psql: timeout expired
repmgr 19:25:45.09 DEBUG ==> Host 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432' is not accessible
psql: timeout expired
repmgr 19:26:05.14 DEBUG ==> Host 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432' is not accessible
psql: timeout expired
repmgr 19:26:25.16 DEBUG ==> Host 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432' is not accessible
psql: timeout expired
repmgr 19:26:45.18 DEBUG ==> Host 'nungpostgres-postgresql-ha-postgresql-0.nungpostgres-postgresql-ha-postgresql-headless.compliance.svc.cluster.local:5432' is not accessible

**Comment by javsalgar** [comment]
I've been trying to reproduce it but I couldn't. I will continue trying to see if I get to the same error.

**Comment by javsalgar** [comment]
I've tried to reproduce it but I don't get the error. Let's see if other users with the issue provide more logs that can help to debug the issue correctly.

**Comment by patrickleet** [comment]
thanks for looking into it

**Comment by davenemeth** [comment]
I have exactly the same issue. I deployed a 2 node cluster and it has been working fine for about a week. Then it suddenly died and can't start again.

The primary tries to connect to the upstream node. It fails with an error, because the that node is not answering on postgres port, since it is waiting for the primary before starting postgres.
So are basically in a deadlock.

**Logs of the primary (node-0):**

postgresql-repmgr 11:23:06.37 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 11:23:06.37 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 11:23:06.37 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 11:23:06.37
postgresql-repmgr 11:23:06.39 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
repmgr 11:23:06.47 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql 11:23:06.47 INFO  ==> Validating settings in POSTGRESQL_* env vars..
repmgr 11:23:06.48 INFO  ==> Querying all partner nodes for common upstream node...
repmgr 11:23:06.52 INFO  ==> There are no nodes with primary role. Assuming the primary role...
repmgr 11:23:06.53 INFO  ==> Preparing PostgreSQL configuration...
postgresql 11:23:06.53 INFO  ==> postgresql.conf file not detected. Generating it...
repmgr 11:23:06.60 INFO  ==> Preparing repmgr configuration...
repmgr 11:23:06.61 INFO  ==> Initializing Repmgr...
postgresql 11:23:06.61 INFO  ==> Initializing PostgreSQL database...
postgresql 11:23:06.61 INFO  ==> Cleaning stale /bitnami/postgresql/data/postmaster.pid file
postgresql 11:23:06.63 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql 11:23:06.63 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql 11:23:06.65 INFO  ==> Deploying PostgreSQL with persisted data...
postgresql 11:23:06.67 INFO  ==> Stopping PostgreSQL...

postgresql-repmgr 11:23:06.68 INFO  ==> ** PostgreSQL with Replication Manager setup finished! **
postgresql 11:23:06.78 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 11:23:09.53 INFO  ==> ** Starting repmgrd **
[2020-06-15 11:23:09] [NOTICE] repmgrd (repmgrd 5.1.0) starting up
INFO:  set_repmgrd_pid(): provided pidfile is /opt/bitnami/repmgr/tmp/repmgr.pid
[2020-06-15 11:23:09] [NOTICE] starting monitoring of node "postgresql-ha-postgresql-0" (ID: 1000)
[2020-06-15 11:23:09] [ERROR] connection to database failed
[2020-06-15 11:23:09] [DETAIL]
could not translate host name "postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.default.svc.cluster.local" to address: Name or service not known

[2020-06-15 11:23:09] [DETAIL] attempted to connect using:
  user=repmgr password=repmgr connect_timeout=5 dbname=repmgr host=postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.default.svc.cluster.local port=5432 fallback_application_name=repmgr
[2020-06-15 11:23:09] [ERROR] unable connect to upstream node (ID: 1001), terminating
[2020-06-15 11:23:09] [HINT] upstream node must be running before repmgrd can start

**Logs of the slave (node-1):**

postgresql-repmgr 11:23:09.04 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 11:23:09.04 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 11:23:09.04 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 11:23:09.04
postgresql-repmgr 11:23:09.06 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
repmgr 11:23:09.13 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql 11:23:09.14 INFO  ==> Validating settings in POSTGRESQL_* env vars..
repmgr 11:23:09.14 INFO  ==> Querying all partner nodes for common upstream node...
repmgr 11:23:09.18 INFO  ==> Preparing PostgreSQL configuration...
postgresql 11:23:09.19 INFO  ==> postgresql.conf file not detected. Generating it...
repmgr 11:23:09.25 INFO  ==> Preparing repmgr configuration...
repmgr 11:23:09.26 INFO  ==> Initializing Repmgr...
repmgr 11:23:09.26 INFO  ==> Waiting for primary node...
postgresql 11:24:19.40 INFO  ==> Stopping PostgreSQL...

**Comment by davenemeth** [comment]
Some additional info to this use case. I try to get it running in Kubernetes which makes it a little more tricky as I see. 
I investigated the chart and found that there is readiness probe, which tries to connect locally to postgres. The pod never gets into ready status, because the startup process is blocked in `repmgr 11:23:09.26 INFO ==> Waiting for primary node...`
The internal load balancer of Kubernetes will not route traffic to an unhealthy pod, so the master will never manage to communicate with the standby. 
I "solved it" by disabling the readiness probe of the standby node, and changed the podmanagementpolicy of the stateful set from default OrderedReady to Parallel.

Now the master can reach the standby, but as there is nobody answering on the postgres port, it results in another error:
`[2020-06-15 13:35:19] [ERROR] connection to database failed
[2020-06-15 13:35:19] [DETAIL]
could not connect to server: Connection refused
        Is the server running on host "postgresql-ha-postgresql-1.postgresql-ha-postgresql-headless.default.svc.cluster.local" (10.42.239.134) and accepting
        TCP/IP connections on port 5432?
`

**Comment by javsalgar** [comment]
Hi,

Even though I cannot reproduce it, it seems to me that we should document or provide an easy way to avoid these deadlocks. Maybe disabling the probes or allowing the service to publish unready addresses may do the trick. You say that, after being able to reach node 1, the daemon is not serving there. Let me see if I understand correctly. Now the primary node can reach the standby, and the standby is not serving on port 5432 because it is waiting for the other node as well?

**Comment by davenemeth** [comment]
Hi,
I have spent quite a lot of time yestarday with this issue and found some more details in the source code. There are a multiple issues playing together which makes it a really nasty problem. 
The deadlock is caused because the containers were killed by Kubernetes in a state where the standby node already knew that there is a new primary, but the primary was not yet promoted, or the process was done properly for some reason. 

After starting up cluster, the following happens:
- the standby waits for the primary and blocks the startup process. This prevents postgres to start, so the primary will never get into ready status (this can be solved by disabling or redefining the readiness probe) and it will not answer to calls on port 5432.
- the primary is in recovery mode and tries to connect to its last known upstream node, which is now already in standby mode and will never answer this call. This leads to the deadlock. 

Without Kubernetes, one could fix it manually with repmgr. This is also not ideal since we want to have a high availability database with automatic failover and no manual interaction at all. Because of Kubernetes/containerization the container gets killed aftrer the repmgr throws the error (unable to connect to upstream) and we have no chance to execute any commands that could fix the problem.

I think this whole situation would have an extremely low chance to happen on a stand alone installation. So Kubernetes is just making it worse, but unfortunately we must get it running on Kubernetes somehow.

**Comment by javsalgar** [comment]
I remember having a similar case with MariaDB galera, and in that context we were allowing to force the recovery to a specific node. Would it make sense to allow a way to force a node to be the new master (skipping the initialization process) so all the nodes would connect to it?

**Comment by davenemeth** [comment]
I am afraid that would not solve the problem in a reliable way. I mean sure, we could force a node to be the new master, but as I understand that would be a manual operation and I would have to know for sure that the node is really the new master. Otherwise I would kill my cluster, I guess. 

The only reliable solution that I currently see would require a change in repmgr. They should change either the primary or the standby startup process. The easiest way would be if the standby node would start postgres and answer calls while it is waiting for the master come up. I am not sure though, if that would cause other problems.

**Comment by javsalgar** [comment]
Hi,

Thanks for the output, I will forward this to the team to see if we can investigate this deadlock issue in more detail. If only there was an easy way to reproduce the issue, we would iterate faster.

**Comment by rafariossaa** [comment]
I was able to reproduce the issue, following these steps:

- Run docker-compose up and wait until it finishes and everything is working as expected. 
- Running an client you will get:
```
I have no name!@4c074b4c9ed6:/$ PGPASSWORD=adminpassword psql -U postgres -h pgpool
psql (11.8)
Type "help" for help.

postgres=# show pool_nodes;
 node_id | hostname | port | status | lb_weight |  role   | select_cnt | load_balance_node | replication_delay | replication_state | replication_sync_state | last_status_change  
---------+----------+------+--------+-----------+---------+------------+-------------------+-------------------+-------------------+------------------------+---------------------
 0       | pg-0     | 5432 | up     | 0.500000  | primary | 0          | false             | 0                 |                   |                        | 2020-07-02 14:14:36
 1       | pg-1     | 5432 | up     | 0.500000  | standby | 0          | true              | 0                 |                   |                        | 2020-07-02 14:14:36
(2 rows)
```
Which means everything is ok.

- Now, stop pg-0 `docker-compose stop pg-0` and after a while, in the logs  `pg-1` will be promoted to `primary` and a few seconds later, pgpool with `choose` pg-1 as `primary`:
```
...
pg-1_1    | [2020-07-02 14:17:56] [NOTICE] monitoring cluster primary "pg-1" (ID: 1001)

...
pgpool_1  | 2020-07-02 14:18:05: pid 140: LOG:  health check retrying on DB node: 0 (round:5)
pgpool_1  | 2020-07-02 14:18:05: pid 140: WARNING:  failed to connect to PostgreSQL server, getaddrinfo() failed with error "Name or service not known"
...
pgpool_1  | 2020-07-02 14:18:24: pid 1: LOG:  execute command: echo ">>> Failover - that will initialize new primary node search!"
pgpool_1  | >>> Failover - that will initialize new primary node search!
pgpool_1  | 2020-07-02 14:18:24: pid 1: LOG:  find_primary_node_repeatedly: waiting for finding a primary node
pgpool_1  | 2020-07-02 14:18:24: pid 1: LOG:  find_primary_node: primary node is 1
pgpool_1  | 2020-07-02 14:18:24: pid 1: LOG:  failover: set new primary node: 1
pgpool_1  | 2020-07-02 14:18:24: pid 1: LOG:  failover: set new master node: 1
```
- Press CTRL-C to stop the compose.
- Now, start again the compose: `docker-compose up`
- The log is as usual, but pgpool will show this message forever:
```
...
pgpool_1  | 2020-07-02 14:37:54: pid 1: LOG:  find_primary_node: standby node is 1
pgpool_1  | 2020-07-02 14:37:55: pid 1: LOG:  find_primary_node: standby node is 1
...
```
- If we get the status of the pool
```
$ docker exec -it lock_client_1 bash
I have no name!@4c074b4c9ed6:/$ PGPASSWORD=adminpassword psql -U postgres -h pgpool
psql (11.8)
Type "help" for help.

postgres=# show pool_nodes;
 node_id | hostname | port | status | lb_weight |  role   | select_cnt | load_balance_node | replicat
ion_delay | replication_state | replication_sync_state | last_status_change  
---------+----------+------+--------+-----------+---------+------------+-------------------+-------------------+-------------------+------------------------+---------------------
 0       | pg-0     | 5432 | down   | 0.500000  | standby | 0          | false             | 0                 |                   |                        | 2020-07-02 14:18:43
 1       | pg-1     | 5432 | up     | 0.500000  | standby | 0          | true              | 0                 |                   |                        | 2020-07-02 14:18:43
(2 rows)
```

**Comment by rafariossaa** [comment]
Hi, 
  We have released new versions of the containers which I think solve this issue, could you give it a try ?

**Comment by websi96** [comment]
Hello everyone,
any progress on this issue? :)

**Comment by websi96** [comment]
I just resolved it by manually changing the repmgrConfiguration of the not starting replica.

In my case i had two replicas:
- postgres-postgresql-ha-postgresql-0
  - standby
  - node_id = 1000
- postgres-postgresql-ha-postgresql-1
  - primary
  - node_id = 1001

I changed my replicacount to 1 and set the remaining one to have node_id 1001 and its name to the second replica, but I left the conninfo to point to the first replica.
With this I tricked my standby replica to think it is the primary and after it started, i manually changed everything in the repmgr database back so that replica 0 is the primary. After this I reverted the custom repmgrConfiguration and everything worked just fine!

Here is the part of the config.yaml:
'''
postgresql:
   repmgrConfiguration: |
       node_id = '1001'
       node_name = 'postgres-postgresql-ha-postgresql-1'
       data_directory = '/opt/bitnami/postgresql/data'
       conninfo = 'user=repmgr password=_pwhere_ host=postgres-postgresql-ha-postgresql-0.postgres-postgresql-ha-postgresql-headless.postgres.svc.cluster.local dbname=repmgr port=5432 connect_timeout=5'
'''

Hope I could help anyone in need and look forward seeing an automatically resolving release :)

**Comment by websi96** [comment]
I further noticed, that this happens, if u are doing a rolling upgrade of your deployment.

**Comment by rafariossaa** [comment]
Hi @websi96 ,
  Thanks for sharing your workaround. 
  Did this also happens to you with the latest images released ?.

**Comment by websi96** [comment]
Hi @rafariossaa ,
Yes, I did an upgrade from 1.4.4 to the latest release.
The upgrade went all good, but then I wanted to configure ldap and when upgrading just the config this happend.

**Comment by rafariossaa** [comment]
Hi, 
  When upgrading you also should indicate the old password, if not a new random one is generated. 
  Do you thinks this could had happened to you ?

**Comment by websi96** [comment]
This could possible have been the problem, but after setting it when I encountered the issue, it didn't fix it by itself..

**Comment by rafariossaa** [comment]
Hi,
  I have created an internal task to look further into this and find a fix.
  I was able to reproduce this with the following steps:
  - Start a cluster
  - Kill the primary node (`node-0`) so other becomes primary (ie: `node-1`).
  - Kill the cluster (helm delete)
  - Start the cluster again (helm install ), in this case the current PVC are reused and also it is needed to indicate the passwords so no new ones are generated.
  The `node-0` is is not able to start because it can not find a primary node. A workaround is the one you indicated before.

  We will come back as soon as we have news.

**Comment by rafariossaa** [comment]
Hi, 
  Could you try the latest version ? I think this issue has been fixed.

**Comment by jgericke** [comment]
FYI - this article provides a fix/workaroud: https://medium.com/@neilcameronwhite/recovering-from-postgresql-ha-failure-on-kubernetes-8c523c986f85

**Comment by rafariossaa** [comment]
Hi @jgericke 
  Thanks for sharing, did you have this issue ? This thread has been stale for a while and I did not get further feedback.

**Comment by jgericke** [comment]
Heys @rafariossaa yeah, we're running 6.5.0 and where hit by this exact deadlock yesterday.

**Comment by carrodher** [comment]
I am not able

**Comment by v0lumehi** [comment]
The issue still occurs. I am running a two node setup in Nomad.
Primary log output:
```
postgresql-repmgr 10:01:27.81 
postgresql-repmgr 10:01:27.81 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 10:01:27.82 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 10:01:27.82 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 10:01:27.82 
postgresql-repmgr 10:01:27.83 DEBUG ==> Configuring libnss_wrapper...
postgresql-repmgr 10:01:27.84 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 10:01:27.87 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 10:01:27.88 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 10:01:27.88 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 10:01:27.89 DEBUG ==> Checking node 'clt1.node.consul:5432'...
psql: error: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
postgresql-repmgr 10:01:27.92 DEBUG ==> Skipping: failed to get primary from the node 'clt1.node.consul:5432'!
postgresql-repmgr 10:01:27.93 INFO  ==> This node was acting as a primary before restart!
postgresql-repmgr 10:01:27.93 INFO  ==> Can not find new primary. Starting PostgreSQL normally...
postgresql-repmgr 10:01:27.94 INFO  ==> There are no nodes with primary role. Assuming the primary role...
postgresql-repmgr 10:01:27.94 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 10:01:27.95 DEBUG ==> Injecting a new postgresql.conf file...
postgresql-repmgr 10:01:27.95 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 10:01:28.06 DEBUG ==> Injecting a new pg_hba.conf file...
postgresql-repmgr 10:01:28.07 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 10:01:28.08 DEBUG ==> Node ID: '1001', Rol: 'primary', Primary Node: ':5432'
postgresql-repmgr 10:01:28.08 INFO  ==> Initializing Repmgr...
postgresql-repmgr 10:01:28.09 INFO  ==> Initializing PostgreSQL database...
postgresql-repmgr 10:01:28.09 INFO  ==> Cleaning stale /bitnami/postgresql/data/postmaster.pid file
postgresql-repmgr 10:01:28.10 DEBUG ==> Copying files from /bitnami/postgresql/conf to /opt/bitnami/postgresql/conf
postgresql-repmgr 10:01:28.10 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/postgresql.conf detected
postgresql-repmgr 10:01:28.10 INFO  ==> Custom configuration /opt/bitnami/postgresql/conf/pg_hba.conf detected
postgresql-repmgr 10:01:28.11 DEBUG ==> Ensuring expected directories/files exist...
postgresql-repmgr 10:01:28.13 INFO  ==> Deploying PostgreSQL with persisted data...
postgresql-repmgr 10:01:28.16 INFO  ==> Configuring replication parameters
postgresql-repmgr 10:01:28.20 INFO  ==> Configuring fsync
postgresql-repmgr 10:01:28.21 DEBUG ==> Skipping repmgr configuration...
postgresql-repmgr 10:01:28.22 INFO  ==> ** PostgreSQL with Replication Manager setup finished! **
postgresql-repmgr 10:01:28.25 INFO  ==> Starting PostgreSQL in background...
postgresql-repmgr 10:01:28.78 INFO  ==> ** Starting repmgrd **
[2021-05-18 10:01:28] [NOTICE] repmgrd (repmgrd 5.2.1) starting up
[2021-05-18 10:01:28] [INFO] connecting to database "user=repmgr password=repmgrpass host=clt1.node.consul dbname=repmgr port=5432 connect_timeout=5"
[2021-05-18 10:01:28] [DEBUG] connecting to: "user=repmgr password=repmgrpass connect_timeout=5 dbname=repmgr host=clt1.node.consul port=5432 fallback_application_name=repmgr options=-csearch_path="
[2021-05-18 10:01:28] [DEBUG] node id is 1001, upstream node id is 1001
INFO:  set_repmgrd_pid(): provided pidfile is /opt/bitnami/repmgr/tmp/repmgr.pid
[2021-05-18 10:01:28] [NOTICE] starting monitoring of node "postgres-metabase-secondary-1" (ID: 1001)
[2021-05-18 10:01:28] [INFO] "connection_check_type" set to "ping"
[2021-05-18 10:01:28] [DEBUG] monitor_streaming_standby()
[2021-05-18 10:01:28] [DEBUG] upstream node ID in local node record is 1001
[2021-05-18 10:01:28] [DEBUG] connecting to upstream node 1001: "user=repmgr password=repmgrpass host=clt3.node.consul dbname=repmgr port=5432 connect_timeout=5"
[2021-05-18 10:01:28] [DEBUG] connecting to: "user=repmgr password=repmgrpass connect_timeout=5 dbname=repmgr host=clt3.node.consul port=5432 fallback_application_name=repmgr options=-csearch_path="
[2021-05-18 10:01:28] [ERROR] connection to database failed
[2021-05-18 10:01:28] [DETAIL] 
could not connect to server: Connection refused
	Is the server running on host "clt3.node.consul" (10.0.2.7) and accepting
	TCP/IP connections on port 5432?

[2021-05-18 10:01:28] [DETAIL] attempted to connect using:
  user=repmgr password=repmgrpass connect_timeout=5 dbname=repmgr host=clt3.node.consul port=5432 fallback_application_name=repmgr options=-csearch_path=
[2021-05-18 10:01:28] [ERROR] unable connect to upstream node (ID: 1001), terminating
[2021-05-18 10:01:28] [HINT] upstream node must be running before repmgrd can start
[2021-05-18 10:01:28] [INFO] repmgrd terminating...
```

Secondary log output:
```
postgresql-repmgr 10:13:31.57 
postgresql-repmgr 10:13:31.58 Welcome to the Bitnami postgresql-repmgr container
postgresql-repmgr 10:13:31.58 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql-repmgr
postgresql-repmgr 10:13:31.58 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql-repmgr/issues
postgresql-repmgr 10:13:31.59 
postgresql-repmgr 10:13:31.60 DEBUG ==> Configuring libnss_wrapper...
postgresql-repmgr 10:13:31.61 INFO  ==> ** Starting PostgreSQL with Replication Manager setup **
postgresql-repmgr 10:13:31.64 INFO  ==> Validating settings in REPMGR_* env vars...
postgresql-repmgr 10:13:31.65 INFO  ==> Validating settings in POSTGRESQL_* env vars..
postgresql-repmgr 10:13:31.65 INFO  ==> Querying all partner nodes for common upstream node...
postgresql-repmgr 10:13:31.66 DEBUG ==> Checking node 'clt1.node.consul:5432'...
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:13:31.69 DEBUG ==> Skipping: failed to get primary from the node 'clt1.node.consul:5432'!
postgresql-repmgr 10:13:31.70 DEBUG ==> Checking node 'clt3.node.consul:5432'...
psql: error: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
postgresql-repmgr 10:13:31.72 DEBUG ==> Skipping: failed to get primary from the node 'clt3.node.consul:5432'!
postgresql-repmgr 10:13:31.72 DEBUG ==> Primary node: 'clt1.node.consul:5432'
postgresql-repmgr 10:13:31.73 INFO  ==> Preparing PostgreSQL configuration...
postgresql-repmgr 10:13:31.74 DEBUG ==> Injecting a new postgresql.conf file...
postgresql-repmgr 10:13:31.74 INFO  ==> postgresql.conf file not detected. Generating it...
postgresql-repmgr 10:13:31.84 DEBUG ==> Injecting a new pg_hba.conf file...
postgresql-repmgr 10:13:31.85 INFO  ==> Preparing repmgr configuration...
postgresql-repmgr 10:13:31.86 DEBUG ==> Node ID: '1001', Rol: 'standby', Primary Node: 'clt1.node.consul:5432'
postgresql-repmgr 10:13:31.86 INFO  ==> Initializing Repmgr...
postgresql-repmgr 10:13:31.86 INFO  ==> Waiting for primary node...
postgresql-repmgr 10:13:31.87 DEBUG ==> Wait for schema repmgr.repmgr on 'clt1.node.consul:5432', will try 6 times with 10 delay seconds (TIMEOUT=60)
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:13:31.89 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:13:41.93 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:13:51.98 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:14:02.03 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:14:12.07 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:14:22.11 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
psql: error: could not connect to server: Connection refused
	Is the server running on host "clt1.node.consul" (10.0.2.5) and accepting
	TCP/IP connections on port 5432?
postgresql-repmgr 10:14:32.15 DEBUG ==> Host 'clt1.node.consul:5432' is not accessible
```

My job description:
```
job "postgres-metabase" {
  datacenters = ["dc1"]
  group "postgres-metabase-primary" {
    network {
      mode = "host"
      port "postgres_metabase" {
        static       = 5432
        host_network = "private"
      }
    }
    volume "postgres" {
      type      = "host"
      read_only = false
      source    = "postgres"
    }
    constraint {
      operator = "distinct_hosts"
      value    = "true"
    }
    task "postgres-metabase-primary-1" {
      driver = "docker"
      constraint {
        attribute = "${node.unique.name}"
        value     = "clt1"
      }
      volume_mount {
        volume      = "postgres"
        destination = "/bitnami/postgresql"
        read_only   = false
      }
      config {
        image       = "bitnami/postgresql-repmgr:13.1.0"
        ports       = ["postgres_metabase"]
        hostname    = "postgres-metabase-primary-1"
        dns_servers = ["10.0.1.5"]
      }
      template {
        data        = <<EOF
POSTGRESQL_USERNAME=postgres
POSTGRESQL_PASSWORD=postgres
POSTGRESQL_DATABASE=metabase
REPMGR_PARTNER_NODES=clt1.node.consul
REPMGR_NODE_NAME=postgres-metabase-primary-1
REPMGR_NODE_NETWORK_NAME=clt1.node.consul
REPMGR_PRIMARY_HOST=clt1.node.consul
REPMGR_START_OPTIONS=--wait-start 0
REPMGR_LOG_LEVEL=DEBUG
BITNAMI_DEBUG=true
REPMGR_PRIMARY_PORT=5432
REPMGR_PASSWORD=repmgrpass
EOF
        destination = "local/postgres-metabase.env"
        env         = true
      }
      service {
        name = "postgres-metabase-primary-1"
        port = "postgres_metabase"
        check {
          name     = "postgres-metabase-primary-1 running tcp open"
          type     = "tcp"
          protocol = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
      resources {
        cpu    = 500
        memory = 1024
      }
    }
  }
  group "postgres-metabase-secondary" {
    network {
      mode = "host"
      port "postgres_metabase" {
        static       = 5432
        host_network = "private"
      }
    }
    volume "postgres" {
      type      = "host"
      read_only = false
      source    = "postgres"
    }
    constraint {
      operator = "distinct_hosts"
      value    = "true"
    }
    task "postgres-metabase-secondary-1" {
      driver = "docker"
      constraint {
        attribute = "${node.unique.name}"
        value     = "clt3"
      }
      volume_mount {
        volume      = "postgres"
        destination = "/bitnami/postgresql"
        read_only   = false
      }
      config {
        image       = "bitnami/postgresql-repmgr:13.1.0"
        ports       = ["postgres_metabase"]
        hostname    = "postgres-metabase-secondary-1"
        dns_servers = ["10.0.1.5"]
      }
      template {
        data        = <<EOF
POSTGRESQL_POSTGRES_PASSWORD=adminpassword
POSTGRESQL_USERNAME=postgres
POSTGRESQL_PASSWORD=postgres
POSTGRESQL_DATABASE=metabase
REPMGR_PARTNER_NODES=clt1.node.consul,clt3.node.consul:5432
REPMGR_NODE_NAME=postgres-metabase-secondary-1
REPMGR_NODE_NETWORK_NAME=clt3.node.consul
REPMGR_PRIMARY_HOST=clt1.node.consul
REPMGR_PRIMARY_PORT=5432
BITNAMI_DEBUG=true
REPMGR_PASSWORD=repmgrpass
EOF
        destination = "local/postgres-metabase.env"
        env         = true
      }
      service {
        name = "postgres-metabase-secondary-1"
        port = "postgres_metabase"
        check {
          name     = "postgres-metabase-secondary-1 running tcp open"
          type     = "tcp"
          protocol = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
      resources {
        cpu    = 500
        memory = 1024
      }
    }
  }
}
```
The error occurs when restarting the task group. When deleting the persisted data before restarting both containers start without any errors. I don't want to delete my data everytime I restart though.

**Comment by v0lumehi** [comment]
Please reopen this issue

**Comment by v0lumehi** [comment]
I was able to create a workaround. I modified the run.sh and built a custom docker image based on yours:
```
#!/bin/bash

# shellcheck disable=SC1091

set -o errexit
set -o nounset
set -o pipefail
# set -o xtrace # Uncomment this line for debugging purpose

# Load libraries
. /opt/bitnami/scripts/liblog.sh
. /opt/bitnami/scripts/libpostgresql.sh
. /opt/bitnami/scripts/librepmgr.sh

# Load PostgreSQL & repmgr environment variables
. /opt/bitnami/scripts/postgresql-env.sh

readonly repmgr_flags=("--pid-file=$REPMGR_PID_FILE" "-f" "$REPMGR_CONF_FILE" "--daemonize=false")
readonly repmgr_cmd=$(command -v repmgrd)

postgresql_start_bg true
info "** Sleeping for 20 seconds **"
sleep 20
info "** Starting repmgrd **"
# TODO: properly test running the container as root
if am_i_root; then
    exec gosu "$POSTGRESQL_DAEMON_USER" "$repmgr_cmd" "${repmgr_flags[@]}"
else
    exec "$repmgr_cmd" "${repmgr_flags[@]}"
fi
```
The only thing I did was adding a 20 second sleep. Don't know if thats a great solution though.

**Comment by rafariossaa** [comment]
Hi @v0lumehi ,
  Thanks for sharing your workaround. Instead of `sleep` it would be better to do a while loop checking if postgresql is currently running. You could use `retry_while` for this.
  In any case, I am not very sure if your issue is the same as the others in this thread because other scenarios are kubernetes related.
  I can see that if postgresql take a while to start, it could produce some errors in repmgr. But, I am not very sure on how is this related with your current error: 
```
2021-05-18 10:01:28] [ERROR] connection to database failed
[2021-05-18 10:01:28] [DETAIL] 
could not connect to server: Connection refused
	Is the server running on host "clt3.node.consul" (10.0.2.7) and accepting
	TCP/IP connections on port 5432?
```
 as the host clt3.node.consul is none of those two in your job.
  I would like to understand this better before implementing changes.

**Comment by v0lumehi** [comment]
clt3.node.consul is being set in the environment variable REPMGR_NODE_NETWORK_NAME. That’s the DNS name of the host that runs the secondary instance. Nomad allows you to deploy containers on specific hosts.

The issue I have is the same as the one reported here. Recreating both primary and secondary containers leads to them waiting for each other. What happens when using the sleep command: The secondary is able to reach the primary instance because the Postgres port of the primary is open. It clones the data of the primary and starts its Postgres service and opens its port. When the sleep timer runs out, the repmgr on the primary starts and is able to reach its upstream, the secondary instance, because the secondary was able to open its port before. Without the upstream (the secondary instance) being available repmgr exits with a non-zero return code.

**Comment by rafariossaa** [comment]
I am no sure what is the missing variable here. Maybe it is the timing due to database size.
Could you tell me how big is your database ?
I was not able to reproduce the issue using docker-compose.

This is what I did:
- Clone this repo and  then `cd bitnami-docker-postgresql-repmgr/13/debian-10`
- Start the compose: `docker-compose up`
- Wait until both nodes are up an running.
- Exec into `docker exec -it debian-10_pg-0_1 bash`
- Get the status replication:
```
$ PGPASSWORD=$POSTGRESQL_POSTGRES_PASSWORD psql -U postgres
psql (13.3)
Type "help" for help.

postgres=# SELECT * FROM pg_stat_replication;
 pid | usesysid | usename | application_name | client_addr | client_hostname | client_port |         backend_star
t         | backend_xmin |   state   | sent_lsn  | write_lsn | flush_lsn | replay_lsn | write_lag | flush_lag | r
eplay_lag | sync_priority | sync_state |          reply_time           
-----+----------+---------+------------------+-------------+-----------------+-------------+---------------------
----------+--------------+-----------+-----------+-----------+-----------+------------+-----------+-----------+--
----------+---------------+------------+-------------------------------
 306 |    16386 | repmgr  | pg-1             | 172.25.0.2  |                 |       56484 | 2021-05-20 07:58:51.
353259+00 |              | streaming | 0/5000788 | 0/5000788 | 0/5000788 | 0/5000788  |           |           |  
          |             0 | async      | 2021-05-20 08:01:42.133623+00
(1 row)
```
- I can create a table and some test data and got it replicated in the  pg-1 node. No errors where observed.
- Stop the compose: `docker-compose down` (without `-v` to preserve data)
- Start again the compose: `docker-compose up`, I can see in the logs the persisted data were used in `pg-0` :
```
pg-0_1  | postgresql-repmgr 08:02:28.29 INFO  ==> Deploying PostgreSQL with persisted data...
```
and node `pg-1`, clone the data from `pg-0`. There was not errors in the logs and the system are up and running.

**Comment by Parthi10** [comment]
Hi, I'm trying to implement postgres-repl in nomad, but primary and replication container's are not communicating.
can you please anyone help me.

Here is my job config,

job "job" {
  datacenters = ["dc1"]
  type        = "service"

  group "group" {

    count = 1

    network {
      mode = "host"
      port "db" {
        static = 20555
        to     = 5432
      }
      port "db1" {
        static = 29052
        to     = 5432
      }
    }

    task "master" {
      kill_timeout = "120s"
      driver       = "docker"
      config {
        image = "bitnami/postgresql-repmgr:latest"
        ports = ["db"]
      }
      env {
        POSTGRESQL_POSTGRES_PASSWORD = "adminpassword"
        POSTGRESQL_USERNAME          = "customuser"
        POSTGRESQL_PASSWORD          = "custompassword"
        POSTGRESQL_DATABASE          = "customdatabase"
        REPMGR_PASSWORD              = "repmgrpassword"
        REPMGR_PRIMARY_HOST          = "master-${NOMAD_ALLOC_ID}"
        REPMGR_PRIMARY_PORT          = "5432"
        REPMGR_PARTNER_NODES         = "naster-${NOMAD_ALLOC_ID}, slave-${NOMAD_ALLOC_ID}:5432"
        REPMGR_NODE_NAME             = "master-${NOMAD_ALLOC_ID}"
        REPMGR_NODE_NETWORK_NAME     = "master-${NOMAD_ALLOC_ID}"
        REPMGR_PORT_NUMBER           = "5432"
        REPMGR_NODE_ID               = "master-${NOMAD_ALLOC_ID}"
      }
      resources {
        memory = 256
      }
    }

    task "slave" {
      driver       = "docker"
      kill_timeout = "120s"
      config {
        image = "bitnami/postgresql-repmgr:latest"
        ports = ["db1"]
      }
      env {
        POSTGRESQL_POSTGRES_PASSWORD  = "adminpassword"
        POSTGRESQL_USERNAME           = "customuser"
        POSTGRESQL_PASSWORD           = "custompassword"
        POSTGRESQL_DATABASE           = "customdatabase"
        REPMGR_PASSWORD               = "repmgrpassword"
        REPMGR_PRIMARY_HOST           = "master-${NOMAD_ALLOC_ID}"
        REPMGR_PRIMARY_PORT           = "5432"
        REPMGR_PARTNER_NODES          = "master-${NOMAD_ALLOC_ID}, slave-${NOMAD_ALLOC_ID}:5432"
        REPMGR_NODE_NAME              = "slave-${NOMAD_ALLOC_ID}"
        REPMGR_NODE_NETWORK_NAME      = "slave-${NOMAD_ALLOC_ID}"
        REPMGR_PORT_NUMBER            = "5432"
        REPMGR_NODE_ID                = "master-${NOMAD_ALLOC_ID}"
      }
      resources {
        memory = 128
      }
    }
  }
}

**Comment by rafariossaa** [comment]
Hi @Parthi10 ,
  Do you mind opening a new issue ? That way we can track it in a more proper way.
  Please, feel free to reference any other issues you consider.
